---
header-includes: |
  \usepackage{gensymb}
  \usepackage[left]{lineno}
  <!-- \linenumbers -->
  \raggedright
  \usepackage[backend=biber,style=authoryear]{biblatex}
  \setlength{\parskip}{1.2\baselineskip}
  \usepackage{pdflscape}
  \newcommand{\blandscape}{\begin{landscape}}
  \newcommand{\elandscape}{\end{landscape}}
  \usepackage{graphicx}
  \usepackage{array}
  \usepackage{booktabs}
  \usepackage{caption}
  \renewcommand{\arraystretch}{0.5}
  \setlength{\tabcolsep}{2pt}
  \AtBeginEnvironment{tabular}{\tiny}
  \AtBeginEnvironment{longtable}{\scriptsize}
  <!-- \widowpenalties 1 10000 -->
  <!-- \raggedbottom -->
classoption:
  - 12pt
geometry:
  - top=25mm
  - left=25mm
linestretch: 2
indent: true
colorlinks: urlcolor, citecolor, toccolor, linkcolor
lof: false
lot: false
thanks: false
toc: false
chapters: True
chaptersDepth: 1
chapDelim: ""
figureTitle: |
  Figure
figPrefix:
  - "Figure"
  - "Figures"
tableTitle: |
  Table
tblPrefix:
  - "Table"
  - "Tables"
showDetails: false
---

<!-- https://github.com/jgm/pandoc/issues/3148 -->

# Title

Author Last and first names

Address

\newpage

# Abstract {-}

Content of abstract.

\newpage

# Introduction {-}

- `BWA` [@Li2009]
- `Minimap2` [@Li2018]
- `SPAdes` [@Bankevich2012]
- `Flye` [@Kolmogorov2019]
- `FMLRC` [@Wang2018]
- `GeSeq` [@Tillich2017]
- `GetOrganelle` [@Jin2020]
- `ptGAUL` [@Zhou2023]
- `CLAW` [@Phillips2024]
- `PMAT` [@Bi2024]
- `Oatk` [@Zhou2024]
- `TIPPo` [@Xian2025]
- Subsampling [@Efron1987]
- `NextDenovo` [@Hu2024]

# Materials and Methods

(@tbl:benchmark-data-some-2)

(@fig:mainfigure1)

- `Flye` [@Kolmogorov2019]
- `JellyFish` [@Marcais2011]
- `BLAST` [@Altschul1997]
- `SeqKit` [@Shen2016]
- `MAFFT` [@Katoh2013]

(@fig:mainfigure1)

- `Bandage` [@Wick2015]

- `Canu` [@Koren2017]
- `NextDenovo` better than `Canu` [@Wick2021]

(@tbl:benchmark-polap-some-0)

(@tbl:benchmark-computer)

# Results

## Comparison with other plastid assembly pipelines

(@tbl:benchmark-polap-some-1)

(Supporting Materials â€“ Plastid genome assemblies using the six pipelines)

(@tbl:benchmark-time-2)

(@tbl:benchmark-memory-2)

## Subsampling-based plastome assemblies

(@tbl:benchmark-polap-some-1)

(@tbl:benchmark-polap-some-0)

(@tbl:benchmark-polap-some-4)

## Three-stage of subsampling-based assembly

- (@tbl:polap-disassemble-Eucalyptus_pauciflora-2)

- (@tbl:polap-disassemble-Eucalyptus_pauciflora-0)

- (@tbl:polap-disassemble-Eucalyptus_pauciflora-3)

# Discussion

`Polap` (Plant Organelle Long-read Assembly Pipeline v0.4.3.7), which includes the subsampling-based plastid genome assembly feature, is available under the GNU General Public License version 3.0 at [http://github.com/goshng/polap](http://github.com/goshng/polap).

# Supplementary Material

Supplementary material, including 10 tables and three figures, is appended to the main text of this manuscript. A `BASH` script for executing the pipeline used to generate the results presented in the manuscript is also included.

# Acknowledgements {-}

We thank Jeffrey L. Thorne for improving the presentation of this work.

# Author Contributions

S.C.C. developed the Polap pipeline and prepared the manuscript.

# Conflict of Interest

The author declare no conflicts.

# Data availability

`Polap` (Plant Organelle Long-read Assembly Pipeline v0.4.3.7) is available under the GNU General Public License version 3.0 at [http://github.com/goshng/polap](http://github.com/goshng/polap).
The results presented in this manuscript are available at Figshare: [https://figshare.com/s/ec1cb394870c7727a2d4](https://figshare.com/s/ec1cb394870c7727a2d4).

# References {-}

::: {#refs}
:::

<!-- # Species -->
<!---->
<!-- - _Anthoceros agrestis_ -->
<!-- - _Arabidopsis thaliana_ -->
<!-- - _Canavalia ensiformis_ -->
<!-- - _Cinchona pubescens_ -->
<!-- - _Codonopsis lanceolata_ -->
<!-- - _Cucumis sativus_var_hardwickii_ -->
<!-- - _Dioscorea japonica_ -->
<!-- - _Dunaliella tertiolecta_ -->
<!-- - _Eucalyptus pauciflora_ -->
<!-- - _Euonymus alatus_ -->
<!-- - _Gossypium herbaceum_ -->
<!-- - _Juncus effusus_ -->
<!-- - _Juncus inflexus_ -->
<!-- - _Juncus roemerianus_ -->
<!-- - _Juncus validus_ -->
<!-- - _Leiosporoceros dussii_ -->
<!-- - _Macadamia jansenii_ -->
<!-- - _Musa acuminata_subsp_malaccensis_ -->
<!-- - _Ophrys lutea_ -->
<!-- - _Oryza rufipogon_ -->
<!-- - _Pisum sativum_ -->
<!-- - _Populus x_sibirica_ -->
<!-- - _Prunus mandshurica_ -->
<!-- - _Pterocarpus santalinus_ -->
<!-- - _Solanum lycopersicum_ -->
<!-- - _Spirodela polyrhiza_ -->
<!-- - _Vaccinium vitis-idaea_ -->
<!-- - _Vitis vinifera_ -->
<!-- - _Notothylas orbicularis_ -->
<!-- - _Phaeomegaceros chiloensis_ -->
<!-- \blandscape -->

\newpage

# Tables

<!--table1-->
<!-- polap-data-cflye man table-benchmark test 2 -->

Table: Plastid genome assemblies for 23 plant species datasets using subsampled sequencing data with a maximum subsampling rate of 5% (Run Setting A).
All datasets were downsampled to 10x genome coverage, and Stage 1 included 10 subsampling steps (N).
Depending on the dataset, different maximum sampling rates (P) were used in Stage 1, and different replicate sizes (R) were applied in Stages 2 and 3.
NA represents no assemblies in the subsampling-based method and no comparison available.
{#tbl:benchmark-polap-test-2}

|Species                |P  |N  |R  |Length (ptGAUL)|Length (Polap)|Percent identity|
|----------------------:|--:|--:|--:|--------------:|-------------:|---------------:|
|_Anthoceros agrestis_  |100|10 |5  |160010         |159938        |99.93813        |
|_Eucalyptus pauciflora_|100|10 |5  |159841         |159945        |99.93185        |

<!-- \elandscape -->

<!--table2-->
<!-- polap-data-cflye man table-polap-disassemble Taxon_genus 2 -->

\newpage

Table: Three stages of subsampling-based plastid genome assembly for the _Eucalyptus pauciflora_ dataset with Run Setting A.
The configuration includes an increasing subsample size up to a maximum subsampling rate of 5%, a step size of 10 in Stage 1, 5 replicates in Stages 2 and 3, and a maximum memory limit of 16 GB.
Abbreviations are as follows: 
iteration in each Stage (I), subsampling rate (Rate) and read-coverage threshold (Alpha);
assembly metrics including the number of segments in the assembly (N),
the total length of these segments (L), and
the number of circular genome paths detected (C); and
the draft plastid genome assembly length (Length).
Alpha at Stage 3 is the percent identity values between consecutive indices.
{#tbl:polap-disassemble-Taxon_genus-2}

|Stage|Index|Rate|Alpha |N  |L     |C  |Memory|Time|Length|
|----:|----:|---:|-----:|--:|-----:|--:|-----:|---:|-----:|
|1    |0    |0.05|1.00  |NA |NA    |NA |8     |1m  |NA    |
|1    |1    |0.16|0.25  |1  |118147|0  |8     |1m  |NA    |
|1    |2    |0.26|0.25  |3  |130443|4  |8     |1m  |156251|
|1    |3    |0.37|0.25  |8  |253143|8  |8     |2m  |NA    |
|1    |4    |0.47|0.25  |3  |130676|4  |9     |2m  |156567|
|1    |5    |0.58|0.25  |4  |173403|4  |10    |2m  |156575|
|1    |6    |0.68|1.00  |5  |132763|4  |12    |2m  |155609|
|1    |7    |0.79|1.75  |3  |130923|4  |10    |2m  |156800|
|1    |8    |0.89|1.00  |5  |272624|4  |16    |2m  |155777|
|2    |0    |0.47|0.25  |6  |265355|4  |9     |3m  |158481|
|2    |1    |0.47|0.25  |3  |132336|4  |9     |2m  |158528|
|2    |2    |0.47|0.25  |4  |180974|4  |9     |2m  |158485|
|2    |3    |0.47|0.25  |4  |193091|4  |9     |2m  |158448|
|2    |4    |0.47|0.25  |5  |247985|4  |9     |2m  |158500|
|3    |0    |0.05|NA    |NA |NA    |NA |.05   |0m  |159338|
|3    |1    |0.29|99.57 |NA |NA    |NA |.12   |.1m |159947|
|3    |2    |0.53|100.00|NA |NA    |NA |.22   |.2m |159944|
|3    |3    |0.76|100.00|NA |NA    |NA |.31   |.3m |159945|
|3    |4    |1.00|100.00|NA |NA    |NA |.40   |.4m |159945|

\newpage

# Figures

<!--figure1-->

![Workflow of the subsampling-based plastid genome assembly. The genome assembly procedure is applied repeatedly in Stages 1 and 2.](figures/polap2-figure1.pdf){#fig:mainfigure1 width=100%}

![Workflow of the subsampling-based plastid genome assembly. The genome assembly procedure is applied repeatedly in Stages 1 and 2.](figures/figure-memory-test-2.pdf){#fig:mainfigure2 width=100%}

![Workflow of the subsampling-based plastid genome assembly. The genome assembly procedure is applied repeatedly in Stages 1 and 2.](figures/figure-time-test-2.pdf){#fig:mainfigure3 width=100%}

\setcounter{table}{0}
\setcounter{figure}{0}
\renewcommand{\thetable}{S\arabic{table}}
\renewcommand{\thefigure}{S\arabic{figure}}

\blandscape

# Supplementary Materials

<!--tableS1-->
<!-- polap-data-cflye man table-benchmark test 2 -->

Table: Sequencing data for the datasets, including species names and their corresponding taxonomic ranks studied. {#tbl:benchmark-data-test-2}

|Species                |Order         |Family         |Long SRA|Long Size|Long Coverage|Short SRA|Short Size|Short Coverage|
|----------------------:|-------------:|--------------:|-------:|--------:|------------:|--------:|---------:|-------------:|
|_Anthoceros agrestis_  |Anthocerotales|Anthocerotaceae|l       |191.5 Mbp|95.58        |s        |191.8 Mbp |95.70         |
|_Eucalyptus pauciflora_|Myrtales      |Myrtaceae      |l       |191.5 Mbp|95.58        |s        |191.8 Mbp |95.70         |

<!-- \begingroup -->
<!-- \tiny -->
<!-- \setlength{\parskip}{2pt} -->
<!---->
<!-- Abbreviations are as follows: -->
<!-- SRA ID for the long-read dataset (L_SRA), -->
<!-- the size of the long-read dataset (L_size), -->
<!-- the long-read coverage (L_cov), -->
<!-- SRA ID for the short-read dataset (S_SRA), -->
<!-- the size of the short-read dataset (S_size), and -->
<!-- the short-read coverage (S_cov). -->
<!---->
<!-- \endgroup -->
<!---->

\elandscape

<!--tableS:computer-->
<!-- polap-data-cflye man table-benchmark test 2 -->

\newpage

Table: Computer setup for the 23 datasets.
{#tbl:benchmark-computer}

|Species                |CPU                 |Cores|Memory|Storage Type|
|----------------------:|-------------------:|----:|-----:|-----------:|
|_Anthoceros agrestis_  |E5-2690 v4 @ 2.60GHz|56   |251Gi |HDD         |
|_Eucalyptus pauciflora_|E5-2690 v4 @ 2.60GHz|56   |251Gi |HDD         |

<!--tableS2-->
<!-- polap-data-cflye man table-benchmark test 1 -->

\newpage

<!-- \blandscape -->

Table: Replicate of plastid genome assemblies for 23 plant species datasets using subsampled sequencing data with a maximum subsampling rate of 5% (Run Setting A).
All datasets were downsampled to 10x genome coverage, and Stage 1 included 10 subsampling steps (N).
Depending on the dataset, different maximum sampling rates (P) were used in Stage 1, and different replicate sizes (R) were applied in Stages 2 and 3.
{#tbl:benchmark-polap-test-1}

|Species                |P  |N  |R  |Length (ptGAUL)|Length (Polap)|Percent identity|
|----------------------:|--:|--:|--:|--------------:|-------------:|---------------:|
|_Anthoceros agrestis_  |100|10 |5  |160010         |159936        |99.93876        |
|_Eucalyptus pauciflora_|100|10 |5  |159841         |159936        |99.93310        |

<!-- \elandscape -->

<!--table:benchmark-time-->
<!-- polap-data-cflye man table-benchmark test 2 -->

\newpage

\blandscape

Table: Benchmark of `GetOrganelle`, `ptGAUL`, `PMAT`, `TIPPo`, `Oatk` and the method (Run Setting A) presented here in terms of data processing time.
NA at the column of `NextDenovo` represents no error-corrected long-read results, resulting in no assemblies in the correction-then-assembly pipelines including `PMAT`, `TIPPo`, and `Oatk`.
Abbreviations are as follows: 
`GetOrganelle` (GO),
`ptGAUL` (pG),
`NextDenovo` (ND),
`PMAT` with `-fc 0.1` (P0.1),
`PMAT` with `-fc 1.0` (P1.0),
`TIPPo` with `-p onthq` (Thq),
`TIPPo` with `-p ont` (Tont),
`Oatk` with `-c 30` (O30), and
`Oatk` with `-c 20` (O20).
{#tbl:benchmark-time-2}

|Species                |GO   |ptG  |MSBWT|FMLRC|ND  |P0.1|P1.0|Thq |Tont|O30|O20|Polap|
|----------------------:|----:|----:|----:|----:|---:|---:|---:|---:|---:|--:|--:|----:|
|_Anthoceros agrestis_  |17.1m|10.0m|2.9m |0m   |4.5m|0m  |1.0m|1.0m|3.0m|0m |0m |1.2h |
|_Eucalyptus pauciflora_|2.7m |1.8m |.3m  |0m   |.9m |0m  |1.0m|1.0m|3.0m|0m |0m |35.1m|

\elandscape

<!--table:benchmark-memory-->
<!-- polap-data-cflye man table-benchmark test 2 -->

\newpage

\blandscape

Table: Benchmark of `GetOrganelle`, `ptGAUL`, `PMAT`, `TIPPo`, `Oatk` and the method (Run Setting A) in terms of peak memory.
NA at the column of `NextDenovo` represents no error-corrected long-read results, resulting in no assemblies in the correction-then-assembly pipelines including `PMAT`, `TIPPo`, and `Oatk`.
Abbreviations are as follows: 
`GetOrganelle` (GO),
`ptGAUL` (pG),
`NextDenovo` (ND),
`PMAT` with `-fc 0.1` (P0.1),
`PMAT` with `-fc 1.0` (P1.0),
`TIPPo` with `-p onthq` (Thq),
`TIPPo` with `-p ont` (Tont),
`Oatk` with `-c 30` (O30), and
`Oatk` with `-c 20` (O20).
{#tbl:benchmark-memory-2}

|Species                |GO  |ptG |MSBWT|FMLRC|ND   |P0.1|P1.0|Thq |Tont|O30 |O20 |Polap|
|----------------------:|---:|---:|----:|----:|----:|---:|---:|---:|---:|---:|---:|----:|
|_Anthoceros agrestis_  |1.87|8.01|0.46 |0.00 |11.76|0.00|2.75|0.66|7.74|0.00|0.00|65.27|
|_Eucalyptus pauciflora_|1.26|7.87|0.18 |0.00 |14.25|0.00|2.69|0.62|7.74|0.00|0.00|15.65|

\elandscape

<!--tableS3-->
<!-- polap-data-cflye man table-benchmark test 0 -->

\newpage

<!-- \blandscape -->

Table: Plastid genome assemblies for 23 plant species datasets using subsampled sequencing data with a maximum subsampling rate of 10% (Run Setting B).
All datasets were downsampled to 10x genome coverage, and Stage 1 included 10 subsampling steps (N).
Depending on the dataset, different maximum sampling rates (P) were used in Stage 1, and different replicate sizes (R) were applied in Stages 2 and 3.
NA represents no assemblies in the subsampling-based method and no comparison available.
{#tbl:benchmark-polap-test-0}

|Species                |P  |N  |R  |Length (ptGAUL)|Length (Polap)|Percent identity|
|----------------------:|--:|--:|--:|--------------:|-------------:|---------------:|
|_Anthoceros agrestis_  |100|10 |5  |160010         |159935        |99.93689        |
|_Eucalyptus pauciflora_|99 |10 |5  |159841         |159932        |99.93560        |

<!-- !include manuscript-table1-footnote.md -->

<!-- \elandscape -->

<!--tableS4-->
<!-- polap-data-cflye man table-benchmark test 4 -->

\newpage

Table: Plastid genome assemblies for 23 plant species datasets using subsampled sequencing data with a maximum subsampling rate of 1%.
All datasets were downsampled to 10x genome coverage, and Stage 1 included 10 subsampling steps (N).
Depending on the dataset, different maximum sampling rates (P) were used in Stage 1, and different replicate sizes (R) were applied in Stages 2 and 3.
NA represents no assemblies in the subsampling-based method and no comparison available.
{#tbl:benchmark-polap-test-4}

|Species                |P  |N  |R  |Length (ptGAUL)|Length (Polap)|Percent identity|
|----------------------:|--:|--:|--:|--------------:|-------------:|---------------:|
|_Anthoceros agrestis_  |100|10 |5  |160010         |159935        |99.94251        |
|_Eucalyptus pauciflora_|100|10 |5  |159841         |159935        |99.93310        |

<!--tableS6-->
<!-- polap-data-cflye man table-polap-disassemble Taxon_genus 0 -->

\newpage

Table: Three stages of subsampling-based plastid genome assembly for the Eucalyptus pauciflora dataset with Run Setting B.
The configuration includes an increasing subsample size up to a maximum subsampling rate of 10%, a step size of 10 in Stage 1, 5 replicates in Stages 2 and 3, and a maximum memory limit of 16 GB.
Abbreviations are as follows: 
iteration in each Stage (I), subsampling rate (Rate) and read-coverage threshold (Alpha);
assembly metrics including the number of segments in the assembly (N),
the total length of these segments (L), and
the number of circular genome paths detected (C); and
the draft plastid genome assembly length (Length).
Alpha at Stage 3 is the percent identity values between consecutive indices.
{#tbl:polap-disassemble-Taxon_genus-0}

|Stage|Index|Rate|Alpha |N  |L     |C  |Memory|Time|Length|
|----:|----:|---:|-----:|--:|-----:|--:|-----:|---:|-----:|
|1    |0    |0.05|1.00  |NA |NA    |NA |8     |1m  |NA    |
|1    |1    |0.16|0.25  |NA |NA    |NA |8     |1m  |NA    |
|1    |2    |0.26|0.25  |1  |153400|2  |8     |1m  |153400|
|1    |3    |0.36|0.25  |3  |129973|4  |8     |1m  |155583|
|1    |4    |0.47|0.25  |3  |130068|4  |9     |1m  |156038|
|1    |5    |0.57|0.25  |3  |130536|4  |9     |2m  |156243|
|1    |6    |0.68|0.25  |3  |130917|4  |12    |2m  |156930|
|1    |7    |0.78|0.25  |5  |170704|4  |12    |2m  |156437|
|1    |8    |0.89|1.00  |4  |208162|4  |13    |3m  |156324|
|1    |9    |0.99|1.75  |3  |129602|4  |14    |2m  |155122|
|2    |0    |0.57|0.25  |4  |159399|4  |9     |2m  |158516|
|2    |1    |0.57|0.25  |3  |132295|4  |9     |2m  |158472|
|2    |2    |0.57|0.25  |2  |197574|2  |9     |2m  |158505|
|2    |3    |0.57|0.25  |4  |214682|4  |9     |2m  |158429|
|2    |4    |0.57|0.25  |3  |132256|4  |9     |2m  |158422|
|3    |0    |0.05|NA    |NA |NA    |NA |.08   |0m  |159223|
|3    |1    |0.29|99.48 |NA |NA    |NA |.12   |.1m |159932|
|3    |2    |0.52|99.99 |NA |NA    |NA |.21   |.2m |159932|
|3    |3    |0.76|100.00|NA |NA    |NA |.31   |.3m |159932|
|3    |4    |0.99|100.00|NA |NA    |NA |.40   |.4m |159934|

<!--tableS7-->
<!-- polap-data-cflye man table-polap-disassemble Taxon_genus 3 -->

\newpage

Table: Three stages of subsampling-based plastid genome assembly for the Eucalyptus pauciflora dataset with Run Setting C.
The configuration includes an increasing subsample size up to a maximum subsampling rate of 10%, a step size of 50 in Stage 1, 5 replicates in Stages 2 and 3, and a maximum memory limit of 16 GB.
Abbreviations are as follows: 
iteration in each Stage (I), subsampling rate (Rate) and read-coverage threshold (Alpha);
assembly metrics including the number of segments in the assembly (N),
the total length of these segments (L), and
the number of circular genome paths detected (C); and
the draft plastid genome assembly length (Length).
Alpha at Stage 3 is the percent identity values between consecutive indices.
{#tbl:polap-disassemble-Taxon_genus-3}

|Stage|Index|Rate|Alpha |N  |L     |C  |Memory|Time|Length|
|----:|----:|---:|-----:|--:|-----:|--:|-----:|---:|-----:|
|1    |0    |0.05|1.00  |NA |NA    |NA |8     |1m  |NA    |
|1    |1    |0.10|0.25  |4  |161519|0  |8     |1m  |NA    |
|1    |2    |0.15|0.25  |10 |261384|8  |8     |1m  |NA    |
|1    |3    |0.20|0.25  |1  |155145|2  |8     |1m  |155145|
|1    |4    |0.25|0.25  |3  |129647|4  |8     |1m  |155380|
|1    |5    |0.30|0.25  |3  |130705|4  |8     |1m  |156598|
|1    |6    |0.35|0.25  |4  |156123|4  |8     |1m  |156499|
|1    |7    |0.40|0.25  |3  |130637|4  |8     |1m  |156583|
|1    |8    |0.45|0.25  |3  |130116|4  |9     |2m  |155977|
|1    |9    |0.50|0.25  |4  |177242|4  |9     |1m  |156759|
|1    |10   |0.55|0.25  |6  |221747|4  |9     |2m  |155102|
|1    |11   |0.60|0.25  |4  |154713|4  |9     |2m  |154635|
|1    |12   |0.65|0.25  |4  |166791|4  |9     |2m  |155916|
|1    |13   |0.70|0.25  |6  |184437|8  |10    |2m  |NA    |
|1    |14   |0.75|0.25  |4  |274222|4  |12    |3m  |156546|
|1    |15   |0.80|1.00  |4  |209428|4  |11    |3m  |155866|
|1    |16   |0.85|1.75  |3  |131009|4  |10    |2m  |156874|
|1    |17   |0.90|1.00  |5  |259272|4  |14    |2m  |156459|
|1    |18   |0.95|1.75  |3  |129146|4  |13    |3m  |153495|
|1    |19   |1.00|2.50  |3  |130928|4  |10    |2m  |156854|
|2    |0    |0.35|0.25  |3  |132312|4  |8     |2m  |158446|
|2    |1    |0.35|0.25  |4  |211664|4  |8     |2m  |158546|
|2    |2    |0.35|0.25  |3  |132387|4  |8     |2m  |158585|
|2    |3    |0.35|0.25  |1  |158527|2  |8     |2m  |158527|
|2    |4    |0.35|0.25  |3  |132021|4  |8     |2m  |158176|
|3    |0    |0.05|NA    |NA |NA    |NA |.05   |0m  |159377|
|3    |1    |0.29|99.58 |NA |NA    |NA |.12   |.1m |159942|
|3    |2    |0.53|100.00|NA |NA    |NA |.22   |.2m |159938|
|3    |3    |0.76|100.00|NA |NA    |NA |.31   |.3m |159938|
|3    |4    |1.00|100.00|NA |NA    |NA |.40   |.4m |159939|

<!--figureS1-->
<!-- polap-data-cflye man figure-alpha Taxon_genus -->

\newpage

![Line plot of read-coverage thresholds versus subsample size index in Stage 1 of the subsampling-based assemblies for _Eucalyptus pauciflora_.](figures/alpha0.pdf){#fig:alpha0 width=100%}

<!--figureS2-->
<!-- polap-data-cflye man figure-delta Taxon_genus -->

\newpage

![Line plot of increment size versus subsample size index in Stage 1 of the subsampling-based assemblies for _Eucalyptus pauciflora_.](figures/delta.pdf){#fig:delta width=100%}

<!--figureS3-->

<!-- Supporting Material - plastid genome assemblies using the six pipelines -->

\newpage

# Code

`Polap` (Plant Organelle Long-read Assembly Pipeline v0.4.3.7) is available at [http://github.com/goshng/polap](http://github.com/goshng/polap).
A quick start guide of the subsampling-based plastome assembly is provided for use on a Linux system with an Internet connection.
A detailed guide is also available, offering a step-by-step explanation of test of the procedures outlined in the quick start.

### Requirements

- **Operating System**: Linux (not compatible with macOS or Windows)
- **Dependencies**: Requires [Bash](https://www.gnu.org/software/bash/) (>= 5.0) and **[Miniconda](https://docs.anaconda.com/miniconda/miniconda-install/)**

### Quick Start

To replicate the results presented in this manuscript on a Linux computer with `git` installed and an Internet connection, follow the steps below. Most steps complete in a relatively short time, except for the final step, which includes both data downloading and full analysis:

```bash
mkdir -p all/polap/cflye1
cd all/polap/cflye1
git clone https://github.com/goshng/polap.git
bash polap/src/polap-data-cflye -y install conda
```

Log out and back in to the terminal.

```bash
cd all/polap/cflye1
source ~/miniconda3/bin/activate
bash polap/src/polap-data-cflye setup conda
bash polap/src/polap-data-cflye -y install minimal
bash polap/src/polap-data-cflye setup polap
```
Log out and back in to the terminal.
```bash
conda activate polap
polap-data-cflye delete-polap-github
polap-data-cflye sample-csv polap-data-v2.csv test
polap-data-cflye -y download-test-data
# run time: about 1 hour
polap-data-cflye local-batch Taxon_genus t off
polap-data-cflye -y install-getorganelle
# polap-data-cflye -y download-pmat
# polap-data-cflye -y install-pmat
polap-data-cflye sample-csv polap-data-v2.csv all on
# edit the CSV file if necessary
polap-data-cflye local-batch each
```

Now, go to step 10 of the next subsection to create tables and figures.

### Detailed Guide

**1. Open a new terminal**:
Open a new terminal in a Linux computer, such as one with Ubuntu.

**2. Install Miniconda**:
Download and install **[Miniconda](https://docs.anaconda.com/miniconda/miniconda-install/)** using the [instructions](https://docs.anaconda.com/miniconda/#quick-command-line-install).
The following is a script that works at the time of writing this manuscript. Otherwise, one could easily find a resource for the installation.

```bash
mkdir -p ~/miniconda3
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh \
  -O ~/miniconda3/miniconda.sh
bash ~/miniconda3/miniconda.sh -b -u -p ~/miniconda3
rm ~/miniconda3/miniconda.sh
```

After installing, close and reopen your terminal application.

**3. Setup the conda channels**:
If you did not close and reopen a new terminal, please do so. Then, execute the followings to setup the conda channels for `polap`.

```bash
source ~/miniconda3/bin/activate
conda config --add channels bioconda
conda config --add channels conda-forge
conda config --set channel_priority strict
```

**4. Install the Bioconda Polap package**:
You setup `polap` and `polap-fmlrc` conda environments using [Polap](https://anaconda.org/bioconda/polap) conda package.

```bash
conda create -y --name polap polap=0.4.3.7.6
```

**5. Installation of Flye for disjointig filtering**:
Note that `Flye` with disjointig filtering feature is a slightly modidified version of the original `Flye`.
You activate the `polap` conda environment and setup `polap-fmlrc` environment

```bash
conda activate polap
conda install -y goshng::cflye
base_dir=$(dirname "$(command -v polap)") && \
  conda env create -f $base_dir/polap-conda-environment-fmlrc.yaml
```

**6. Polap assemble run with a test dataset**:
This tests the basic execution of the `polap` commmand.

```bash
wget -q https://github.com/goshng/polap/archive/refs/tags/0.4.3.7.6.zip
unzip -o -q 0.4.3.7.6.zip
cd polap-0.4.3.7.6/test
polap assemble --test
```

**7. Plastid genome assembly with _Eucalyptus pauciflora_ dataset**:
Your assembled plastid genome sequence will be `o/ptdna.0.fa`.

```bash
polap x-ncbi-fetch-sra --sra SRR7153095
polap x-ncbi-fetch-sra --sra SRR7161123
polap disassemble -l SRR7153095.fastq \
  -a SRR7161123_1.fastq \
  -b SRR7161123_2.fastq
```

**8. Check the accuracy of the plastid genome assembly**:
We use the Polap disassemble command with _Eucalyptus pauciflora_ dataset and check its similarity with its known plastid genome sequence
Your assembled plastid genome sequence will be `o/ptdna.ref.0.fa`. The text file named `o/0/mafft/pident.txt` has the percent identity between the assembled ptDNA and the knomn reference.

```bash
polap get-mtdna --plastid --species "Eucalyptus pauciflora"
cp o/00-bioproject/2-mtdna.fasta o/ptdna-reference.fa
polap disassemble \
  --disassemble-i 1 \
  --stages-include 3 \
  -l SRR7153095.fastq \
  -a SRR7161123_1.fastq \
  -b SRR7161123_2.fastq \
  --disassemble-align-reference \
  --disassemble-c o/ptdna-reference.fa

mkdir -p o/0/mafft
polap mafft-mtdna \
  -a o/ptdna-reference.fa \
  -b o/0/disassemble/2/pt.subsample-polishing.reference.aligned.1.fa \
  -o o/0/mafft
cat o/0/mafft/pident.txt
```

**9. Batch script that creates the results in the manuscript**:

```bash
polap-data-cflye -y install-getorganelle
# polap-data-cflye -y download-pmat
# polap-data-cflye -y install-pmat
polap-data-cflye example-data polap-data-v2.csv all on
polap-data-cflye local-batch each
```

**10. Tables in the manuscript**:
Tables in Markdown format will be generated and saved in the `man` directory after executing the following command.
You should download a precompiled binary version 0.8.1 of `Bandage` genome assembly graph visualization tool from [the official Bandage GitHub](https://github.com/rrwick/Bandage/releases).

```bash
polap-data-cflye -y install-bandage
# Install xelatex if necessary ...
# sudo apt-get install texlive texlive-latex-recommended texlive-xetex
# sudo apt-get install texlive-fonts-recommended texlive-fonts-extra texlive-lang-all
polap-data-cflye -y install-man
polap-data-cflye -y download-man
polap-batch-v2.sh
polap-data-cflye -y make-man
```

<!-- benchmark assembly graph figure -->
<!-- polap-data-cflye man figure-sheet test 2 -->
